upup/pkg/fi/cloudup/awstasks/load_balancer_cleanup.go

```diff
@@ -0,0 +1,461 @@
+
+func (d *detachNetworkLoadBalancer) Item() string {
+	tmp := *d.request.TargetGroupARNs[0] + " -> " + *d.request.AutoScalingGroupName
+	return aws.StringValue(&tmp)
+}
+
+func (d *detachNetworkLoadBalancer) Delete(t fi.Target) error {
+	klog.V(2).Infof("deleting elb %v", d)
+
+	awsTarget, ok := t.(*awsup.AWSAPITarget)
+	if !ok {
+		return fmt.Errorf("unexpected target type for deletion: %T", t)
+	}
+
+	name := aws.StringValue(d.request.AutoScalingGroupName)
+	klog.V(2).Infof("Calling autoscaling Detach LoadBalancer for autoscaling group %s", name)
+
+	_, err := awsTarget.Cloud.Autoscaling().DetachLoadBalancerTargetGroups(d.request)
+
+	if err != nil {
+		return fmt.Errorf("Error Detaching LoadBalancer TargetGroup from Autoscaling group : %v", err)
+	}
+
+	return nil
+}
+
+func (d *detachNetworkLoadBalancer) String() string {
+	return d.TaskName() + "-" + d.Item()
+}
+
+func (e *LoadBalancerCleanup) FindELBDeletions(c *fi.Context) ([]fi.Deletion, error) {
+	var removals []fi.Deletion
+
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	lb, err := FindLoadBalancerByNameTag(cloud, fi.StringValue(e.LBName))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if lb != nil {
+
+		request := &elb.DeleteLoadBalancerInput{
+			LoadBalancerName: lb.LoadBalancerName,
+		}
+
+		removals = append(removals, &deleteLoadBalancer{request: request})
+		klog.V(2).Infof("will delete load balancer: %v", lb.LoadBalancerName)
+	}
+
+	for _, autoScalingGroupName := range e.AgNames {
+
+		request := &autoscaling.DescribeLoadBalancersInput{
+			AutoScalingGroupName: autoScalingGroupName,
+		}
+		response, err := cloud.Autoscaling().DescribeLoadBalancers(request)
+
+		if err != nil || response == nil { //TODO: Mock Autoscaling is returning nil
+			return nil, nil
+		}
+
+		for _, LoadBalancerState := range response.LoadBalancers { //detach all elbs from autoscaling group
+
+			loadBalancerName := LoadBalancerState.LoadBalancerName
+
+			request := &autoscaling.DetachLoadBalancersInput{
+				AutoScalingGroupName: autoScalingGroupName,
+				LoadBalancerNames: []*string{
+					loadBalancerName,
+				},
+			}
+
+			removals = append(removals, &detachLoadBalancer{request: request})
+			klog.V(2).Infof("will detach load balancer: %v from autoscalinggroup %v", loadBalancerName, autoScalingGroupName)
+		}
+	}
+
+	return removals, nil
+}
+
+func (e *LoadBalancerCleanup) FindNLBDeletions(c *fi.Context) ([]fi.Deletion, error) {
+	var removals []fi.Deletion
+
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	lb, err := FindNetworkLoadBalancerByNameTag(cloud, fi.StringValue(e.LBName))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if lb != nil {
+
+		request := &elbv2.DeleteLoadBalancerInput{
+			LoadBalancerArn: lb.LoadBalancerArn,
+		}
+
+		removals = append(removals, &deleteNetworkLoadBalancer{request: request})
+		klog.V(2).Infof("will delete network load balancer: %v", lb.LoadBalancerName)
+
+		//TODO: Could be useful to check the autoscaling group for associated target groups and delete them as well.
```

Comment:
the InstanceGroupSpec supports providing additional target group ARNs to be attached to the ASG, so we would need to make sure we're not deleting the attachment of one of those. I don't think that's necessary though, so we can probably remove this TODO